internal/authentication/ldap_user_provider.go

```diff
@@ -209,6 +209,88 @@ func (p *LDAPUserProvider) UpdatePassword(username, password string) (err error)
 	return nil
 }
 
+// UpdatePassword update the password of the given user.
+func (p *LDAPUserProvider) ChangePassword(username, oldPassword string, newPassword string) (err error) {
+	var (
+		client  LDAPClient
+		profile *ldapUserProfile
+	)
+
+	if client, err = p.connect(); err != nil {
+		return fmt.Errorf("unable to update password for user '%s'. Cause: %w", username, err)
+	}
+
+	defer client.Close()
+
+	if profile, err = p.getUserProfile(client, username); err != nil {
+		return fmt.Errorf("unable to update password for user '%s'. Cause: %w", username, err)
+	}
+
+	var controls []ldap.Control
+
+	switch {
+	case p.features.ControlTypes.MsftPwdPolHints:
+		controls = append(controls, &controlMsftServerPolicyHints{ldapOIDControlMsftServerPolicyHints})
+	case p.features.ControlTypes.MsftPwdPolHintsDeprecated:
+		controls = append(controls, &controlMsftServerPolicyHints{ldapOIDControlMsftServerPolicyHintsDeprecated})
+	}
+
+	userPasswordOk, err := p.CheckUserPassword(username, oldPassword)
+
+	if err != nil {
+		if strings.Contains(err.Error(), ErrIncorrectPassword.Error()) {
+			userPasswordOk = false
+		} else {
+			return err
+		}
+	}
```

Comment:
We should probably avoid this method of detection and instead look to use the LDAP result codes. I know we have done it similar to this in the past but using the way defined in the spec is a method we can guarantee will work with the majority of implementations.

See here: https://github.com/go-ldap/ldap/blob/master/v3/error.go#L38

Here's an example:

```go
func errToChangePasswordError(target error) (err error) {
	var e *ldap.Error

	if errors.As(err, &e) && e.ResultCode == ldap.LDAPResultInvalidCredentials {
		return ErrIncorrectPassword
	}

	return target
}
```